File: python/entity/record.py
"""
entity/record.py — data models mirroring the Go entity package.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Dict


@dataclass
class Record:
    """A simple record: an integer ID with a string-to-string data map."""

    id: int
    data: Dict[str, str] = field(default_factory=dict)

    def copy(self) -> "Record":
        """Return a shallow copy so callers cannot mutate stored state."""
        return Record(id=self.id, data=dict(self.data))

    def to_dict(self) -> dict:
        return {"id": self.id, "data": self.data}


@dataclass
class VersionedRecord:
    """A Record annotated with its version number and creation timestamp."""

    id: int
    version: int
    data: Dict[str, str]
    created_at: datetime

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "version": self.version,
            "data": self.data,
            "created_at": self.created_at.strftime("%Y-%m-%dT%H:%M:%SZ"),
        }

File: entity/record.go
package entity

import "time"

type Record struct {
	ID   int               `json:"id"`
	Data map[string]string `json:"data"`
}

func (d *Record) Copy() Record {
	values := d.Data

	newMap := map[string]string{}
	for key, value := range values {
		newMap[key] = value
	}

	return Record{
		ID:   d.ID,
		Data: newMap,
	}
}

// VersionedRecord represents a record at a specific point in time.
type VersionedRecord struct {
	ID        int               `json:"id"`
	Version   int               `json:"version"`
	Data      map[string]string `json:"data"`
	CreatedAt time.Time         `json:"created_at"`
}
